explorer/interpreter/interpreter.cpp

```diff
@@ -204,7 +208,11 @@ class Interpreter {
 //
 
 void Interpreter::TraceState() {
-  *trace_stream_ << "{\nstack: " << todo_ << "\nmemory: " << heap_ << "\n}\n";
+  if (stack_heap_sizes_.first != todo_.size() ||
+      stack_heap_sizes_.second != heap_.size()) {
```

Comment:
An action might pop one state and then push another one, which would keep the size the same. Similarly, an action might write to a heap entry without changing the size of the heap. I think some more fine-grained checks would work better here -- let's trace each time the heap or stack changes, when it changes.

How about this:

- For the stack: add `ActionStack::Push` and `ActionStack::Pop` functions, and change all the calls to `todo_.Push` and `todo_.Pop` to call those instead. Trace the new stack after each change to it.
- For the heap: add tracing to `Heap::AllocateValue`, `Heap::Write`, and `Heap::Deallocaate`.

Then remove `Interpreter::TraceState` and all calls to it, as we'll already be tracing the heap and stack each time they change.

Reply:
Thanks for this suggestion. I would implement it this way in the next commit.

Reply:
Thanks for the suggestion! Will do the suggested changes in the next commit.